File: code/py/ch08/08.04.py
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score

my_url = ('https://raw.githubusercontent.com'
          '/taroyabuki/fromzero/master/data/wine.csv')
my_data = pd.read_csv(my_url)

n = len(my_data)
my_data2 = my_data.assign(rand1=np.random.random(n), rand2=np.random.random(n))
my_data2.head()
#>    LPRICE2  WRAIN  DEGREES  HRAIN  TIME_SV     rand1     rand2
#> 0 -0.99868    600  17.1167    160       31  0.089076  0.877133
#> 1 -0.45440    690  16.7333     80       30  0.645344  0.397617
#> 2 -0.80796    502  17.1500    130       28  0.238390  0.238026
#> 3 -1.50926    420  16.1333    110       26  0.753152  0.951262
#> 4 -1.71655    582  16.4167    187       25  0.148555  0.866770

X, y = my_data2.drop(columns=['LPRICE2']), my_data2['LPRICE2']
my_model2 = LinearRegression().fit(X, y)

y_ = my_model2.predict(X)
mean_squared_error(y_, y)**0.5
#> 0.2530051445030606 # RMSE（訓練）

my_scores = cross_val_score(my_model2, X, y,
                            cv=len(y),                        # LOOCV           
                            scoring='neg_mean_squared_error') # MSE
(-my_scores.mean())**0.5
#> 0.35191104236501053 # RMSE（検証）

File: code/R/ch08/08.04.R
## 8.4 入力変数の数とモデルの良さ

library(caret)
library(tidyverse)
my_url <- str_c("https://raw.githubusercontent.com",
                "/taroyabuki/fromzero/master/data/wine.csv")
my_data <- read_csv(my_url)

n <- nrow(my_data)
my_data2 <- my_data %>% mutate(v1 = 0:(n - 1) %% 2,
                               v2 = 0:(n - 1) %% 3)
head(my_data2)
#> # A tibble: 6 x 7
#>   LPRICE2 WRAIN DEGREES HRAIN TIME_SV    v1    v2
#>     <dbl> <dbl>   <dbl> <dbl>   <dbl> <dbl> <dbl>
#> 1  -0.999   600    17.1   160      31     0     0
#> 2  -0.454   690    16.7    80      30     1     1
#> 3  -0.808   502    17.2   130      28     0     2
#> 4  -1.51    420    16.1   110      26     1     0
#> 5  -1.72    582    16.4   187      25     0     1
#> 6  -0.418   485    17.5   187      24     1     2

my_model2 <- train(form = LPRICE2 ~ ., data = my_data2, method = "lm",
                   trControl = trainControl(method = "LOOCV"))
y  <- my_data2$LPRICE2
y_ <- my_model2 %>% predict(my_data2)

RMSE(y_, y)
#> [1] 0.256212 # RMSE（訓練）

my_model2$results$RMSE
#> [1] 0.3569918 # RMSE（検証）

